src/react-integration/provider/middleware.ts

```diff
@@ -12,28 +11,62 @@ export default function createEnhancedReducerHook(
   const useEnhancedReducer = <R extends React.Reducer<any, any>>(
     reducer: R,
     startingState: React.ReducerState<R>,
-  ): [React.ReducerState<R>, React.Dispatch<React.ReducerAction<R>>] => {
+  ): [
+    React.ReducerState<R>,
+    (value: React.ReducerAction<R>) => Promise<any>,
+  ] => {
+    const stateRef = useRef(startingState);
+    const dispatchPromiseRef = useRef<null | PromiseHolder>(null);
     const [state, realDispatch] = useReducer(reducer, startingState);
-    const store = useRef(state);
-    store.current = state;
+
+    useEffect(() => {
+      stateRef.current = state;
+      if (dispatchPromiseRef.current) {
+        dispatchPromiseRef.current.resolve();
+        dispatchPromiseRef.current = null;
+      }
+    }, [state]);
```

Comment:
I'm not sure this will work.

Because this effect is ran when `state` is updated, and the # of times `state` is updated doesn't necessarily equal the number of times `dispatch` is called due to batched state changes, you'll end up with some `dispatch` promises that never get resolved.

To get around this, you can wrap the `dispatch` function so it adds a promise to the action sent to the reducer and returns that promise, then wrap the reducer so it resolves that action with the new state after it's run.

Reply:
Batching works fine because it only creates new promises when one doesn't exist. It will reuse the existing promise for all dispatches until the next commit.

Reply:
So here's a concern I have - imagine the following scenario

1. function `f` calls dispatch, attaches a callback to the promise.
2. In the same event loop iteration, function `k` calls dispatch, attaches a callback to the promise.
3. `f`'s state change is flushed,
4. The state is now updated for `f`, but `f` and `k` have the same promise, so `k`'s promise must be resolved, but `k`s update hasn't happened yet.